lib/util/hud_code_gen.rb

```diff
@@ -0,0 +1,150 @@
+###
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module HudCodeGen
+  CODEGEN_FILE_HEADER = '# THIS FILE IS GENERATED, DO NOT EDIT DIRECTLY'.freeze
+
+  MAP_NAME_OVERRIDES = {
+    export_period_types: :period_types,
+    no_yes_missings: :yes_no_missing_options,
+    no_yes_reasons_for_missing_data: :no_yes_reasons_for_missing_data_options,
+    name_data_qualities: :name_data_quality_options,
+    ssn_data_qualities: :ssn_data_quality_options,
+    dob_data_qualities: :dob_data_quality_options,
+    residence_prior_length_of_stays: :length_of_stays,
+    times_homeless_past_three_years: :times_homeless_options,
+    months_homeless_past_three_years: :month_categories,
+    relationship_to_ho_hs: :relationships_to_hoh,
+    when_dv_occurreds: :when_occurreds,
+    event_types: :events,
+    dependent_under6s: :dependent_under_6,
+    hopwa_financial_assistances: :hopwa_financial_assistance_options,
+    ssvf_financial_assistances: :ssvf_financial_assistance_options,
+    path_referrals: :path_referral_options,
+    rhy_referrals: :rhy_referral_options,
+    bed_nights: :bed_night_options,
+    voucher_trackings: :voucher_tracking_options,
+    moving_on_assistances: :moving_on_assistance_options,
+  }.stringify_keys.freeze
+
+  LOOKUP_FN_OVERRIDES = {
+    relationship_to_ho_h: :relationship_to_hoh,
+    when_dv_occurred: :when_d_v_occurred,
+    event_type: :event,
+    dependent_under6: :dependent_under_6,
+  }.stringify_keys.freeze
+
+  module_function
+
+  def generate_hud_lists(year = '2024')
+    source = File.read("lib/data/#{year}_hud_lists.json")
+    all_lists = JSON.parse(source).sort_by { |hash| hash['code'] }
+    skipped = []
+    filename = "lib/util/concerns/hud_lists_#{year}.rb"
+    arr = []
+    arr.push ::Code.copywright_header
+    arr.push "
+      # frozen_string_literal: true
+      #{CODEGEN_FILE_HEADER}
+      module Concerns::HudLists#{year}
+        extend ActiveSupport::Concern
+          class_methods do
+            def _translate(map, id, reverse)
+              if reverse
+                rx = forgiving_regex id
+                if rx.is_a?(Regexp)
+                  map.detect { |_, v| v.match?(rx) }.try(&:first)
+                else
+                  map.detect { |_, v| v == rx }.try(&:first)
+                end
+              else
+                map[id] || id
+              end
+            end
```

Comment:
This is fine here, but I wonder about just putting it in `hud_validation_util.rb` since it's consistent.  It is nice that it's in the file where it gets used, but also feels kinda odd since it doesn't really have anything to do directly with the lists.

Reply:
Yeah, sure. It is weird having it here.